labrad/fasttypes.c

```diff
@@ -0,0 +1,2364 @@
+}
+
+int ft_write_to_buf(COBJ *cobj, char *buf) {
+    unsigned int ui;
+    int i, n, size = 0;
+    
+    if (cobj->type == EMPTY) {
+        size = 0;
+    }
+    else if (cobj->type == BOOL) {
+        size = LABRAD_BOOL_SIZE;
+        if (cobj->i) *buf = TRUE;
+        else *buf = FALSE;
+    }
+    else if (cobj->type == UINT31 || cobj->type == INT32) {
+        size = LABRAD_INT_SIZE;
+        ft_appropriate_memcpy(buf, &(cobj->i), size);
+    }
+    else if (cobj->type == UINT32) {
+        size = LABRAD_INT_SIZE;
+        ft_appropriate_memcpy(buf, &(cobj->n), size);
+    }
+    else if (cobj->type == FLOAT64 || cobj->type == FLOAT64_SU) {
+        size = LABRAD_FLOAT_SIZE;
+        ft_appropriate_memcpy(buf, &(cobj->x), size);
+    }
+    else if (cobj->type == COMPLEX128 || cobj->type == COMPLEX128_SU) {
+        size = 2*LABRAD_FLOAT_SIZE;
+        ft_appropriate_memcpy(buf, &(cobj->x), LABRAD_FLOAT_SIZE);
+        ft_appropriate_memcpy(buf + LABRAD_FLOAT_SIZE, &(cobj->y), LABRAD_FLOAT_SIZE);
+    }
+    else if (cobj->type == STRING) {
+        size = LABRAD_INT_SIZE + cobj->n;
+        ft_appropriate_memcpy(buf, &(cobj->n), LABRAD_INT_SIZE);
+        memcpy(buf + LABRAD_INT_SIZE, cobj->str, cobj->n);
+    }
+    else if (cobj->type == LIST || cobj->type == CLUSTER) {
+        size = LABRAD_INT_SIZE;
+        ft_appropriate_memcpy(buf, &(cobj->n), LABRAD_INT_SIZE);
+        for (ui=0; ui<cobj->n; ui++) {
+            size += ft_write_to_buf(cobj->cobj_arr[ui], buf+size);
+        }
+    }
+    else if (cobj->type == NDARRAY || cobj->type == NDARRAY_SU) {
+        size = 0;
+        n = 1;
+        for (ui=0; ui<cobj->n; ui++) {
+            ft_appropriate_memcpy(buf+size, &(cobj->dim_arr[ui]), LABRAD_INT_SIZE);
+            size += LABRAD_INT_SIZE;
+            n *= cobj->dim_arr[ui];
+        }
+        if (cobj->subtype == BOOL) {
+            memcpy(buf+size, cobj->arr, n*LABRAD_BOOL_SIZE);
+        }
+        else if (cobj->subtype == INT32) {
+            for (i=0; i<n; i++) {
+                ft_appropriate_memcpy(buf+size, &((int *)cobj->arr)[i], LABRAD_INT_SIZE);
+                size += LABRAD_INT_SIZE;
+            }
+        }
+        else if (cobj->subtype == UINT32) {
+            for (i=0; i<n; i++) {
+                ft_appropriate_memcpy(buf+size, &((unsigned int *)cobj->arr)[i], LABRAD_INT_SIZE);
+                size += LABRAD_INT_SIZE;
+            }
+        }
+        else if (cobj->subtype == FLOAT64) {
+            for (i=0; i<n; i++) {
+                ft_appropriate_memcpy(buf+size, &((double *)cobj->arr)[i], LABRAD_FLOAT_SIZE);
+                size += LABRAD_FLOAT_SIZE;
+            }
+        }
+        else if (cobj->subtype == COMPLEX128) {
+            for (i=0; i<2*n; i++) {
+                ft_appropriate_memcpy(buf+size, &((double *)cobj->arr)[i], LABRAD_FLOAT_SIZE);
+                size += LABRAD_FLOAT_SIZE;
+            }
+        }
+    }
+    
+    return size;
+}
+
+void ft_appropriate_m
```

Comment:
Thanks for the improvements!
On Apr 22, 2015 10:05 PM, "ejeffrey" <notifications@github.com> wrote:

> In labrad/fasttypes.c
> <https://github.com/martinisgroup/pylabrad/pull/88#discussion_r28936730>:
>
> > +                ft_appropriate_memcpy(buf+size, &((double *)cobj->arr)[i], LABRAD_FLOAT_SIZE);
> > +                size += LABRAD_FLOAT_SIZE;
> > +            }
> > +        }
> > +        else if (cobj->subtype == COMPLEX128) {
> > +            for (i=0; i<2*n; i++) {
> > +                ft_appropriate_memcpy(buf+size, &((double *)cobj->arr)[i], LABRAD_FLOAT_SIZE);
> > +                size += LABRAD_FLOAT_SIZE;
> > +            }
> > +        }
> > +    }
> > +
> > +    return size;
> > +}
> > +
> > +void ft_appropriate_m
>
> tc_list needs an initializer, other wise if line 2303 causes an exception,
> it will be use uninitialized in the exception handler.
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/martinisgroup/pylabrad/pull/88/files#r28936730>.
>